query-engine/core/src/query_graph_builder/extractors/filters/mod.rs

```diff
@@ -144,7 +144,133 @@ pub fn extract_filter(value_map: ParsedInputMap, model: &ModelRef) -> QueryGraph
             _ => Ok(Filter::and(filters)),
         }
     }
-    extract_filter(value_map, model, 0)
+
+    let filter = extract_filter(value_map, model, 0)?;
+    let filter = merge_search_filters(filter);
+
+    Ok(filter)
+}
+
+/// Search filters that have the same query and that are in the same condition block
+/// are merged together to optimize the generated SQL statements.
+/// This is done in three steps (below transformations are using pseudo-code):
+/// 1. We flatten the filter tree.
+/// eg: `Filter(And([ScalarFilter, ScalarFilter], And([ScalarFilter])))` -> `Filter(And([ScalarFilter, ScalarFilter, ScalarFilter]))`
```

Comment:
This step is generally useful, we may want to pull that up into the filter processing.

Reply:
I'm always afraid of doing that, because if for whatever reason the flattening is removed, we end up with a function 3 layers below that no longer receives the expected data. And the link between the flattening and the folding of search filters is invisible.

I'd really be interested in knowing how you usually deal with these situations. For now, given that you're off and that we can't discuss things, I'll keep it that way.